include/unix/osd.h

```diff
@@ -51,6 +51,73 @@
 #define OFI_UNUSED UNREFERENCED_PARAMETER
 #endif
 
+/* Atomic memory access macro */
+static inline
+long ofi_atomic_inc32(volatile long *ptr)
+{
+	return __sync_fetch_and_add(ptr, (long)1);
+}
+
+static inline
+long long ofi_atomic_inc64(volatile long long *ptr)
+{
+	return __sync_fetch_and_add(ptr, (long long)1);
+}
+
+static inline
+long ofi_atomic_dec32(volatile long *ptr)
+{
+	return __sync_fetch_and_sub(ptr, (long)1);
+}
+
+static inline
+long long ofi_atomic_dec64(volatile long long *ptr)
+{
+	return __sync_fetch_and_sub(ptr, (long long)1);
+}
+
+static inline
+long ofi_atomic_add32(volatile long *ptr, long value)
+{
+	return __sync_fetch_and_add(ptr, (long)value);
+}
+
+static inline
+long long ofi_atomic_add64(volatile long long *ptr, long long value)
+{
+	return __sync_fetch_and_add(ptr, (long)value);
+}
+
+static inline
+long ofi_atomic_sub32(volatile long *ptr, long value)
+{
+	return __sync_fetch_and_sub(ptr, (long)value);
+}
+
+static inline
+long long ofi_atomic_sub64(volatile long long *ptr, long long value)
+{
+	return __sync_fetch_and_sub(ptr, (long long)value);
+}
+
+static inline
+void* ofi_val_compare_and_swap_ptr(volatile void **ptr, void *ptr1, void *ptr2)
+{
+    return (void *)__sync_val_compare_and_swap(ptr, ptr1, ptr2);
+}
+
+static inline
+long ofi_val_compare_and_swap32(volatile long *ptr, long oldval, long newval)
+{
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
+
+static inline
+long long ofi_val_compare_and_swap64(volatile long long *ptr, long oldval, long newval)
+{
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
```

Comment:
There are atomic functions defined in include/fi_atom.h.  Those should be used, modified, or added to as needed.  That code handles situations where native atomic support is not available.  We may need to restructure the code there to handle windows.  It may be best to submit this patch in its own PR, so we can cross platform check it.